modules/optimization/src/utils/ParameterMesh.C

```diff
@@ -14,22 +14,86 @@
+                             const std::string & exodus_mesh,
+                             const std::vector<std::string> & var_names)
   : _communicator(MPI_COMM_SELF), _mesh(_communicator)
 {
   _mesh.allow_renumbering(false);
-  _mesh.read(exodus_mesh);
-
+  _mesh.prepare_for_use();
+  _exodusII_io = std::make_unique<ExodusII_IO>(_mesh);
+  _exodusII_io->read(exodus_mesh);
+  _mesh.read(exodus_mesh); // fixme lynn,  I think I still need this
   // Create system to store parameter values
   _eq = std::make_unique<EquationSystems>(_mesh);
   _sys = &_eq->add_system<ExplicitSystem>("_parameter_mesh_sys");
   _sys->add_variable("_parameter_mesh_var", param_type);
-  _eq->init();
 
   // Create point locator
   _point_locator = PointLocatorBase::build(TREE_LOCAL_ELEMENTS, _mesh);
   _point_locator->enable_out_of_mesh_mode();
+
+  if (!var_names.empty())
+  {
+    // Make Exodus vars for equation system
+    const std::vector<std::string> & all_nodal(_exodusII_io->get_nodal_var_names());
+    const std::vector<std::string> & all_elemental(_exodusII_io->get_elem_var_names());
+
+    std::vector<std::string> nodal_variables;
+    std::vector<std::string> elemental_variables;
+    for (const auto & var_name : var_names)
+    {
+      if (std::find(all_nodal.begin(), all_nodal.end(), var_name) != all_nodal.end())
+        nodal_variables.push_back(var_name);
+      if (std::find(all_elemental.begin(), all_elemental.end(), var_name) != all_elemental.end())
+        elemental_variables.push_back(var_name);
+    }
+    if ((elemental_variables.size() + nodal_variables.size()) != var_names.size())
+    {
+      std::string out("\n  Parameter Group Variables Requested: ");
+      for (const auto & var : var_names)
+        out += var + " ";
+      out += "\n  Exodus Nodal Variables: ";
+      for (const auto & var : all_nodal)
+        out += var + " ";
+      out += "\n  Exodus Elemental Variables: ";
+      for (const auto & var : all_elemental)
+        out += var + " ";
+      mooseError("Exodus file did not contain all of the parameter names being intitialized.", out);
+    }
+
+    if (!nodal_variables.empty() && !elemental_variables.empty())
+    {
+      std::string out("\n  Parameter Group Nodal Variables: ");
+      for (const auto & var : nodal_variables)
+        out += var + " ";
+      out += "\n  Parameter Group Elemental Variables: ";
+      for (const auto & var : elemental_variables)
+        out += var + " ";
+      mooseError("Parameter group contains nodal and elemental variables, this is "
+                 "not allowed.  ",
+                 out);
+    }
+    // Add the parameter group ics and bounds to the system
+    // All parameters in a group will be the same type, only one of these loops will do anything
+    if (!nodal_variables.empty())
+      for (const auto & var_name : nodal_variables)
+        _sys->add_variable(var_name, param_type);
+    else if (!elemental_variables.empty())
```

Comment:
```suggestion
      for (const auto & var_name : nodal_variables)
        _sys->add_variable(var_name, param_type);
```
for loop will adapt to empty vectors

Reply:
ok!  For some reason I thought I might have both types of variables but that won't be the case.